### PR TITLE
[SPARK-31432][DEPLOY] Make SPARK_JARS_DIR configurable

### PR DESCRIPTION
**Pull request by marblejenka**
### What changes were proposed in this pull request?

This PR makes SPARK_JARS_DIR configurable in scripts under bin/sbin and launcher.

### Why are the changes needed?

As I wrote in SPARK-31432, the change is benefitical for flexibility of deployment.

### Does this PR introduce any user-facing change?

No, as long as the user dose not set SPARK_JARS_DIR in the environemnt variable.

### How was this patch tested?

scripts under bin/sbin
- I built release archive by ./dev/make-distribution.sh and run some tests by hand on linux and windows. The log of the test is https://gist.github.com/marblejenka/28cb3e14056e72c8d6e540c999516459

launcher
- I have added a testcase assuming SPARK_JARS_DIR is not set
- It is possible to write a testsase for SPARK_JARS_DIR is set, but it is difficult to do it with portability. I will add the testcase if reviewer requests.

### CODE REVIEW

**Comment by maropu** [comment]
ok to test

**Comment by maropu** [comment]
cc: @dongjoon-hyun @srowen

File: launcher/src/main/java/org/apache/spark/launcher/CommandBuilderUtils.java [hunk 1]
```diff
@@ -328,4 +329,9 @@ static String findJarsDir(String sparkHome, String scalaVersion, boolean failIfN
     return libdir.getAbsolutePath();
   }
 
+  static String getJarsDir(String sparkHome) {
+    String jarsDir = System.getenv(ENV_SPARK_JARS_DIR);
+    return jarsDir != null ? jarsDir : join(File.separator, sparkHome, "jars");
+  }
+
```
Comment (maropu) [inline launcher/src/main/java/org/apache/spark/launcher/CommandBuilderUtils.java:336]: nit: remove the blank.

File: launcher/src/test/java/org/apache/spark/launcher/CommandBuilderUtilsSuite.java [hunk 2]
```diff
@@ -22,6 +22,7 @@
 import java.util.List;
 
 import org.junit.Test;
+
```
Comment (maropu) [inline launcher/src/test/java/org/apache/spark/launcher/CommandBuilderUtilsSuite.java:25]: nit: unnecessary change?

**Comment by planga82** [comment]
I think it is very useful, I'm going to open a pull request to introduce it in the spark configuration documentation.
Thanks!

**Review by srowen — APPROVED** [review]

File: launcher/src/test/java/org/apache/spark/launcher/CommandBuilderUtilsSuite.java (hunk 2 above)
Comment (marblejenka) [reply launcher/src/test/java/org/apache/spark/launcher/CommandBuilderUtilsSuite.java:25]: Yes, I overlooked it. I will remove it.

File: launcher/src/main/java/org/apache/spark/launcher/CommandBuilderUtils.java (hunk 1 above)
Comment (marblejenka) [reply launcher/src/main/java/org/apache/spark/launcher/CommandBuilderUtils.java:336]: This is because most of the code in launcher put a blank line after the last method. Shoud I fix it?

Comment (srowen) [reply launcher/src/main/java/org/apache/spark/launcher/CommandBuilderUtils.java:336]: It's trivial either way. I personally prefer a blank line here as other methods are followed with a blank line.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by dongjoon-hyun** [comment]
Thank you for pinging me, @maropu .

File: docs/configuration.md [hunk 3]
```diff
@@ -2939,6 +2939,10 @@ To specify a different configuration directory other than the default "SPARK_HOM
 you can set SPARK_CONF_DIR. Spark will use the configuration files (spark-defaults.conf, spark-env.sh, log4j.properties, etc)
 from this directory.
 
+# Overriding classpath default directory
+
+To specify a different classpath directory other than the default "SPARK_HOME/jars"  you can set SPARK_JARS_DIR. Any of the services under "SPARK_HOME/sbin" folder will use it.
```
Comment (dongjoon-hyun) [inline docs/configuration.md:2944]: nit, `"  you` -> `",  you`

Comment (dongjoon-hyun) [reply docs/configuration.md:2944]: Why do we limit to only `sbin` here?

File: launcher/src/test/java/org/apache/spark/launcher/CommandBuilderUtilsSuite.java [hunk 4]
```diff
@@ -99,6 +99,11 @@ public void testJavaMajorVersion() {
     assertEquals(10, javaMajorVersion("10"));
   }
 
+  @Test
+  public void testGetJarsDir() {
+    assertEquals("/spark/home/jars", getJarsDir("/spark/home"));
```
Comment (dongjoon-hyun) [inline launcher/src/test/java/org/apache/spark/launcher/CommandBuilderUtilsSuite.java:104]: Hi, @marblejenka .
This seems to add a half of the test coverage.
Since this PR's key contribution is to introduce `ENV_SPARK_JARS_DIR`, shall we have a test coverage for `ENV_SPARK_JARS_DIR`, too?

File: docs/configuration.md (hunk 3 above)
Comment (dongjoon-hyun) [reply docs/configuration.md:2944]: Like the PR title have `bin` and `sbin`, this PR changes the standalone ExternalShuffleService like the following. We had better mention that.
```
bin/spark-class org.apache.spark.deploy.ExternalShuffleService
```

**Comment by dongjoon-hyun** [comment]
Thank you for your first contribution, @marblejenka !
I left a few minor comments.

File: launcher/src/main/java/org/apache/spark/launcher/CommandBuilderUtils.java [hunk 5]
```diff
@@ -315,7 +316,7 @@ static int javaMajorVersion(String javaVersion) {
    */
   static String findJarsDir(String sparkHome, String scalaVersion, boolean failIfNotFound) {
     // TODO: change to the correct directory once the assembly build is changed.
-    File libdir = new File(sparkHome, "jars");
+    File libdir = new File(getJarsDir(sparkHome));
     if (!libdir.isDirectory()) {
       libdir = new File(sparkHome, String.format("assembly/target/scala-%s/jars", scalaVersion));
       if (!libdir.isDirectory()) {
```
Comment (dongjoon-hyun) [inline launcher/src/main/java/org/apache/spark/launcher/CommandBuilderUtils.java:322]: Ur, this PR seems to cause a misleading error message when the user give a wrong `SPARK_JARS_DIR`.

In case of a wrong `SPARK_JARS_DIR`, the error message should give the user-given `SPARK_JARS_DIR` instead of `assembly/target/scala-%s/jars`. Could you check this once more, @marblejenka ?

Also, it would be great if we can have a negative test coverage for the above wrong `SPARK_JARS_DIR`.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Although the existing error message scheme seems to show the last tried `libdir`,  this PR seems to add more complexity with `SPARK_JARS_DIR` in terms of the error message.

In case of `findJarsDir()`,
1. Old lookup behavior: `sparkHome/jars` -> `assembly/target/scala-%s/jars`
2. New lookup behavior in this PR: `SPARK_JARS_DIR` -> `assembly/target/scala-%s/jars`

Currently, this PR doesn't complain about the wrong `SPARK_JARS_DIR`. I believe the following is more compatible with the existing behavior (and the existing installation).

3. `SPARK_JARS_DIR` -> `sparkHome/jars` -> `assembly/target/scala-%s/jars`

In addition to (3), if we can identify the user's wrong `SPARK_JARS_DIR`, it would be great if we can give `WARN` message.

**Comment by marblejenka** [comment]
@dongjoon-hyun Thanks for the analysis. Based on your comment, I will work on the following.
- add a test in CommandBuilderUtilsSuite for coverage
- modify the behavior of CommandBuilderUtils#findJarsDir in case SPARK_JARS_DIR is specified but seems not to appropriate

**Comment by marblejenka** [comment]
@dongjoon-hyun I pushed the modified code. Note that I have added a dependency for  com.github.stefanbirkner:system-rules https://github.com/stefanbirkner/system-rules to the testing of the environment variable. Since it is licensed Common Public License Version 1.0, I think it is not a problem for spark project.
I also modified the document based on planga82's work. Thank you @planga82 !

File: pom.xml [hunk 6]
```diff
@@ -928,6 +928,18 @@
         <version>0.11</version>
         <scope>test</scope>
       </dependency>
+      <dependency>
+        <groupId>com.github.stefanbirkner</groupId>
```
Comment (dongjoon-hyun) [inline pom.xml:932]: Ur, @marblejenka . If possible, could you avoid to introduce new test dependency? Is it inevitable?

Comment (marblejenka) [reply pom.xml:932]: It is not impossible, but we have to do the same thing to test code that uses environment variables. Consequently, we will include a copy of that dependency (specifically https://github.com/stefanbirkner/system-rules/blob/master/src/main/java/org/junit/contrib/java/lang/system/EnvironmentVariables.java , which enables to write environment variable via reflection api) in our test code. 
The other way is giving up coverage, but I think it not desirable.

In my opinion, system-rules is beneficial for testing of environment variables in other usecase. Additionally, there should be no problem with the license, so I would like you to merge this.

File: launcher/src/main/java/org/apache/spark/launcher/CommandBuilderUtils.java [hunk 7]
```diff
@@ -314,18 +319,31 @@ static int javaMajorVersion(String javaVersion) {
...
-          libdir.getAbsolutePath());
-        return null;
-      }
+      return handleJarsDir(libdir, failIfNotFound,
+          "Library directory '%s' does not exist; make sure Spark is built.");
     }
     return libdir.getAbsolutePath();
   }
 
+  private static String handleJarsDir(File theDir, boolean failIfNotFound, String message) {
```
Comment (dongjoon-hyun) [inline launcher/src/main/java/org/apache/spark/launcher/CommandBuilderUtils.java:339]: `theDir` -> `dir`.

File: launcher/src/main/java/org/apache/spark/launcher/CommandBuilderUtils.java [hunk 8]
```diff
@@ -314,18 +319,31 @@ static int javaMajorVersion(String javaVersion) {
    * or a distribution directory.
    */
   static String findJarsDir(String sparkHome, String scalaVersion, boolean failIfNotFound) {
+    String envSparkJarsDir = System.getenv(ENV_SPARK_JARS_DIR);
+    if (envSparkJarsDir != null && !envSparkJarsDir.isEmpty()) {
+      File envJarsDir = new File(envSparkJarsDir);
+      return handleJarsDir(envJarsDir, failIfNotFound,
```
Comment (dongjoon-hyun) [inline launcher/src/main/java/org/apache/spark/launcher/CommandBuilderUtils.java:325]: `return`? What I suggested is to warn and continue the old lookup behavior.
```
SPARK_JARS_DIR -> sparkHome/jars -> assembly/target/scala-%s/jars
```

File: pom.xml (hunk 6 above)
Comment (dongjoon-hyun) [reply pom.xml:932]: `environmentVariables.set` is used only twice. I guessed the following will be enough for this PR. What do you think?
```java
    try {
      System.setProperty("SPARK_JARS_DIR", "...");
      // test code
    } finally {
      System.clearProperty("SPARK_JARS_DIR");
    }
```

File: launcher/src/test/java/org/apache/spark/launcher/CommandBuilderUtilsSuite.java [hunk 9]
```diff
@@ -99,6 +111,56 @@ public void testJavaMajorVersion() {
...
+
+    // if we have SPARK_HOME/jars, use it as jars dir
+    tempFolder.newFolder("spark", "home", "jars");
+    assertEquals(
+      join(File.separator, tmpRoot, "spark", "home", "jars"),
+      findJarsDir(sparkHome, "2.13", true));
+
+    // if we have SPARK_JARS_DIR, use it as jars dir
+    File spefifiedJarsDir = tempFolder.newFolder("jars-dir");
+    environmentVariables.set("SPARK_JARS_DIR", spefifiedJarsDir.getAbsolutePath());
```
Comment (dongjoon-hyun) [inline launcher/src/test/java/org/apache/spark/launcher/CommandBuilderUtilsSuite.java:145]: `"SPARK_JARS_DIR"` -> `ENV_SPARK_JARS_DIR`.

File: launcher/src/test/java/org/apache/spark/launcher/CommandBuilderUtilsSuite.java [hunk 10]
```diff
@@ -99,6 +111,56 @@ public void testJavaMajorVersion() {
...
+    assertEquals(join(File.separator, tmpRoot, "jars-dir"),
+      findJarsDir(sparkHome, "2.13", true));
+
+    // if SPARK_JARS_DIR is specified but not exists, throws IllegalStateException
+    environmentVariables.set("SPARK_JARS_DIR", join(File.separator, tmpRoot, "jars-dir-not-exists"));
```
Comment (dongjoon-hyun) [inline launcher/src/test/java/org/apache/spark/launcher/CommandBuilderUtilsSuite.java:150]: `"SPARK_JARS_DIR"` -> `ENV_SPARK_JARS_DIR`.

**Comment by dongjoon-hyun** [comment]
Gentle ping, @marblejenka ~

File: pom.xml (hunk 6 above)
Comment (marblejenka) [reply pom.xml:932]: System#getProperties and System#getenv are slightly different. Since I use the environment variable, we can't use the suggested way(I choose System#getenv because SPARK_CONF_DIR also uses environment variable). 
https://docs.oracle.com/javase/8/docs/api/java/lang/System.html#getProperties--
https://docs.oracle.com/javase/8/docs/api/java/lang/System.html#getenv--

If adding new dependencies in this PR is unacceptable, I will open another PR aimed at improving test coverage using environment variables.
It seems that some code that uses environment variables is not branch-covered according to a coverage report I created. While improving branch coverage of existing stable code may not be very useful, I don't think it is harmful.

File: launcher/src/main/java/org/apache/spark/launcher/CommandBuilderUtils.java (hunk 7 above)
Comment (marblejenka) [reply launcher/src/main/java/org/apache/spark/launcher/CommandBuilderUtils.java:339]: I will fix it. Thank you.

File: launcher/src/main/java/org/apache/spark/launcher/CommandBuilderUtils.java (hunk 8 above)
Comment (marblejenka) [reply launcher/src/main/java/org/apache/spark/launcher/CommandBuilderUtils.java:325]: I will fix it. Thank you.

File: launcher/src/test/java/org/apache/spark/launcher/CommandBuilderUtilsSuite.java (hunk 9 above)
Comment (marblejenka) [reply launcher/src/test/java/org/apache/spark/launcher/CommandBuilderUtilsSuite.java:145]: I will fix it. Thank you.

File: launcher/src/test/java/org/apache/spark/launcher/CommandBuilderUtilsSuite.java (hunk 10 above)
Comment (marblejenka) [reply launcher/src/test/java/org/apache/spark/launcher/CommandBuilderUtilsSuite.java:150]: I will fix it. Thank you.

**Comment by marblejenka** [comment]
I will correct the points you have pointed out.
Sorry for the late response.

**Comment by tgravescs** [comment]
I'm not sure I understand the description of the use case here, you want to point to a separate JAR directory but keep the same SPARK_CONF_DIR?  I'm trying to understand why not set SPARK_HOME to a different location.

note that on yarn there are other options for jars in HDFS like spark.yarn.jars and spark.yarn.archive which will be more efficient using the distributed cache.

**Comment by nchammas** [comment]
One use case I have for this feature is to make up for the fact that `spark.jars` and `spark.driver.extraClassPath` don't support relative paths. This is handy if you use Spark locally to run tests or to execute small-scale data tasks.

For example, say you have two projects on your local machine that each use PySpark.

```
project1/
  spark-conf/
    spark-defaults.conf
    jars/
      some.jar
  local-tool1.py
  requirements.txt
project2/
  spark-conf/
    spark-defaults.conf
    jars/
      another.jar
  local-tool2.py
  requirements.txt
```

You install/manage PySpark for each project via separate Python virtual environments. The projects need different versions of Spark, and depend on different jars and packages that Spark needs to be pointed to. You specify all of that via each project's spark-defaults.conf, and set `SPARK_CONF_DIR` depending on what project you're working on.

The problem is that you also want to point Spark at the appropriate jars, and spark-defaults.conf is not that helpful since you must specify absolute paths for `spark.jars` and `spark.driver.extraClassPath`. The absolute path to the jars will be different for each developer, since they can clone these projects to arbitrary directories, and since the absolute path will likely include their username in some way.

I believe `SPARK_JARS_DIR` solves this problem. We can remove any mention of `spark.jars` or `spark.driver.extraClassPath` from the defaults files, which are checked in to version control, and instead have people set `SPARK_JARS_DIR` depending on what project they are using.

I know this is not the use case that @marblejenka had in mind, but I wanted to share it since we have this problem at my current job. Perhaps this is more an argument for using Spark via Docker instead of virtual environments, but I wanted to share it for the record.

**Comment by marblejenka** [comment]
@tgravescs Certainly, we can do the same thing with SPARK_JARS_DIR by having multiple SPARK_HOME. One of the motivations for this PR is to have a single SPARK_CONF_DIR. Also, regarding spark.yarn.jars and spark.yarn.archive, it does not change the dependency of spark-shell process for example, so it is not sufficient for my usecase.
@nchammas Thank you for sharing your use case. However, I think that SPARK_DIST_CLASSPATH may work. SPARK_DIST_CLASSPATH allows you to add additional jars to your classpath. I'm not a user of pyspark, so there might be some misunderstanding.

Overall, SPARK_JARS_DIR is useful when you want to manage multiple sets of jars.

**Comment by marblejenka** [comment]
@dongjoon-hyun I have fixed the points you have pointed out except for the additional dependency for testing. If adding new dependencies in this PR is unacceptable, I think opening another PR aimed at improving test coverage using environment variables is best so far.

**Comment by tgravescs** [comment]
so normally I would think SPARK_CONF_DIR is separate from your spark home dir anyway. The production envs and hdp deployment I've seen were all separate from SPARK_HOME. Doesn't mean it has to be though.   I can see times it could be useful but other times it could be confusing.  Like someone updates spark to another version and updates the jars but doesn't get the changes for other components like pyspark/R.  But I guess that is a deployment issue.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
Thank you for updating, @marblejenka . 

According to the above discussion, it seems that the scope of this PR should be narrow downed. `spark-shell` is a good environment, but the majority of job submissions is `spark-submit` with `cluster` mode. And, @tgravescs pointed out, `spark.yarn.jars` and `spark.yarn.archive` is the popular way in YARN environment. Could you be more specific in the PR description about **the newly beneficial scope**? Is it beneficially only for `spark-shell`? Also, it would be great if you describe the alternatives like `spark.yarn.jars` and `spark.yarn.archive` together to be fair in the PR description.
> it does not change the dependency of spark-shell process for example, so it is not sufficient for my usecase.

**Comment by marblejenka** [comment]
I understand. Since I'm starting parental leave this week, I think the response will be this weekend.

**Comment by marblejenka** [comment]
Due to family circumstances, I cannot spend time on this issue for a while. I think I can respond in the middle of May. I'm sorry.
